### ISSUE TITLE
`torch.matmul()` with transposed Tensor gives inconsistent behavior

### ISSUE THREAD

**Issue by tht99** [pytorch/pytorch#117739]
### 🐛 Describe the bug

Here is the code to reproduce this weird behavior:
```python
import torch
t1 = torch.rand((1000,768))
x1 = torch.matmul(t1, t1.T) # 1.34 ms ± 42.8 µs per loop
y1 = torch.matmul(t1, t1.T.detach().clone()) # 1.92 ms ± 48.2 µs per loop
assert x1.allclose(y1) # True

t2 = torch.rand((10000,768))
x2 = torch.matmul(t2, t2.T) # 270 ms ± 2.83 ms per loop
y2 = torch.matmul(t2, t2.T.detach().clone()) # 181 ms ± 2.45 ms per loop
assert x2.allclose(y2) # True
```

For the smaller 2D Tensor, `torch.matmul()` with its transposed is faster than the clone. However, the clone version is faster for the bigger 2D Tensor.

I'm not sure why this happened. Shouldn't one way be faster than the other?

### Versions

PyTorch version: 2.1.2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.2.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.7 [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] torch==2.1.2
[conda] numpy                     1.26.3          py311he598dae_0  
[conda] numpy-base                1.26.3          py311hfbfe69c_0  
[conda] torch                     2.1.2                    pypi_0    pypi

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @kulinseth @albanD @DenisVieriu97 @razarmehr

**Comment by malfet** [comment]
operations on GPU are asynchronous, so that's not a valid way to measure the performance of the operation, one needs to measure the time between two `torch.mps.synchronize()` calls

**Comment by malfet** [comment]
There are few factor in play here:
 - Time needed to compute `torch.mm(x, x.T.clone().detach())` consists of matmul + time needed to call `clone` + `detach`, which for small matrices can be longer then running the matmul
 - `.T` operator does not allocate new tensor, just returns a view of the same tensor, which makes [data locality](https://cvw.cac.cornell.edu/code-optimization/data-locality/data-locality-matters) for `torch.mm` much worse compared to running it for matrix multiplication across row first and column first matrices (which `T.clone()` achieves)


See https://pytorch.org/blog/tensor-memory-format-matters/ that tries to talk in depth about why it matters

Please use https://discuss.pytorch.org/ to ask questions about PyTorch, but do not hesitate to create a new issue if you encounter an undefined behavior
